File: algorithms/93_RestoreIPAddresses.py
class Solution:
    def restoreIpAddresses(self, s):
        if len(s) < 4: return []
        res = []
        
        def dfs(remaining, ip, count):
            if count == 4:
                if not remaining:
                    res.append(ip[1:])
                return
            
            bound = min(len(remaining) + 1, 4)
            for i in range(1, bound):
                if remaining[0] == "0" and i > 1:
                    return
                if int(remaining[:i]) < 256:
                    dfs(remaining[i:], ip + "." + remaining[:i], count+1)
        dfs(s, "", 0)
        return res
        

File: algorithms/dfs/93_RestoreIPAddresses.java
class Solution {
    public List<String> restoreIpAddresses(String s) {
        List<String> res = new LinkedList<String>();
        if (s.length() < 4) {
            return res;
        }
        DFS(s, "", 0, res);
        return res;
    }
    
    public void DFS(String remaining, String ip, int count, List<String> l) {
        if (count == 4) {
            if (remaining.length() == 0) {
                l.add(ip.substring(1));
            }
            return;
        }

        int limit = Math.min(remaining.length() + 1, 4);
        for (int i = 1; i < limit; ++i) {
            if (remaining.charAt(0) == '0' && i > 1) {
                return;
            }
            if (Integer.parseInt(remaining.substring(0, i)) < 256) {
                DFS(remaining.substring(i), ip + "." + remaining.substring(0, i), count+1, l);
            }
        }
    }
}